komga-webui/src/functions/reader.ts

```diff
@@ -0,0 +1,396 @@
+  }
+
+  set pageCursor (value: number) {
+    this._pageCursor = value
+  }
+
+  get count (): number {
+    return this.spreads?.spreads?.length!!
+  }
+
+  get progress (): number {
+    return (this.realPageNumber / this.realPageCount) * 100
+  }
+
+  get realPageNumber (): number {
+    return this.spreads.spreads[this.pageCursor]?.pages[0]?.number
+  }
+
+  get realPageCount (): number {
+    return this.reader.pages.length
+  }
+
+  get jump (): Jump {
+    return this._jump
+  }
+
+  goToFirst (): void {
+    this.goTo(1)
+  }
+
+  goToLast (): void {
+    this.goTo(this.count - 1)
+  }
+
+  goTo (page: number): void {
+    this.pageCursor = this.reader?.spreads.indexOf.get(page)!!
+  }
+
+  forward (force: boolean = false) {
+    if (!force && this.isVertical()) return
+    return this.flipDirection() ? this.next() : this.prev()
+  }
+
+  backward (force: boolean = false) {
+    if (!force && this.isVertical()) return
+    return this.flipDirection() ? this.prev() : this.next()
+  }
+
+  up () {
+    if (this.isVertical()) {
+      this.prev()
+    }
+  }
+
+  down () {
+    if (this.isVertical()) {
+      this.next()
+    }
+  }
+
+  vtouch () {
+    let swipe = this.swipe()
+    return {
+      left: () => {
+        if (swipe) {
+          this.forward()
+        }
+      },
+      right: () => {
+        if (swipe) {
+          this.backward()
+        }
+      },
+      up: () => {
+        if (swipe) {
+          this.down()
+        }
+      },
+      down: () => {
+        if (swipe) {
+          this.up()
+        }
+      },
+    }
+  }
+
+  private swipe (): boolean {
+    return this.reader.get(ReaderSettings.SWIPE)
+  }
+
+  private flipDirection (): boolean {
+    return this.reader.get(ReaderSettings.READ_DIR) === ReadingDirection.LEFT_TO_RIGHT
+  }
+
+  private isVertical (): boolean {
+    return this.reader.get(ReaderSettings.READ_DIR) === ReadingDirection.VERTICAL
+  }
+
+  private hasNext () {
+    return this._pageCursor < this.count
+  }
+
+  private hasPrev () {
+    return this._pageCursor >= 1
+  }
+
+  private next () {
+    if (this.hasNext()) {
+      this._pageCursor++
+    } else {
+      if (this.jump.nextBook) {
+        this.nextBook()
+      } else {
+        this.jump.nextBook = true
+      }
+    }
+  }
+
+  private prev () {
+    if (this.hasPrev()) {
+      this._pageCursor--
+    } else {
+      if (this.jump.prevBook) {
+        this.prevBook()
+      } else {
+        this.jump.prevBook = true
+      }
+    }
+  }
+
+  nextBook () {
+    if (this.reader.siblingNext) {
+      this.jump.nextBook = false
+      this.goToBook(this.reader.siblingNext)
+    } else {
+      this.closeBook()
+    }
+  }
+
+  prevBook () {
+    if (this.reader.siblingPrev) {
+      this.jump.prevBook = false
+      this.goToBook(this.reader.siblingPrev)
+    }
+  }
+
+  goToBook (sibling: BookDto) {
+    this.reader.$router.push({ name: 'read-book', params: { bookId: sibling.id.toString()!! } })
+  }
+
+  closeBook () {
```

Comment:
i find it difficult to understand that the reader or navigator is performing actions, it should raise events, and the BookReader should take the actions.